technologydata/sources.py

```diff
@@ -208,6 +222,378 @@ def process(self, trusted_execution: bool = False) -> None:
+            if the expected substrings are not found in the archive URL).
+
+        """
+        archive_url = savepagenow.capture_or_cache(url_to_archive)
+        try:
+            # The timestamp is between "web/" and the next "/" afterward
+            # Find the starting index of "web/"
+            start_index = archive_url[0].index("web/") + len("web/")
+            # Find the ending index of the timestamp by locating the next "/" after the start_index
+            end_index = archive_url[0].index("/", start_index)
+            # Extract the timestamp substring
+            timestamp = archive_url[0][start_index:end_index]
+            output_timestamp = td.Utils.change_datetime_format(
+                timestamp,
+                td.DateFormatEnum.SOURCES_CSV,
+            )
+            return archive_url[0], archive_url[1], output_timestamp
+        except ValueError:
+            # If "web/" or next "/" not found, return empty string
+            return None
+
+    def download_file_from_wayback(self) -> list[pathlib.Path | None]:
+        """
+        Download a file from the Wayback Machine and save it to a specified path.
+
+        This method retrieves an archived file from the Wayback Machine using the URL
+        stored in the `details` attribute of the instance. The file is saved in the
+        specified format based on its Content-Type field in the Response Header.
+        Supported formats include:
+        - Plain text (.txt)
+        - PDF (.pdf)
+        - Excel (.xls and .xlsx)
+        - Parquet (.parquet)
+
+        Returns
+        -------
+        pathlib.Path | None
+            The specified path where the file is stored, or None if an error occurs.
+
+        Raises
+        ------
+        requests.exceptions.RequestException
+            If there is an issue with the HTTP request.
+
+        Notes
+        -----
+        - The `details` attribute must contain a key "url_archived" with a valid URL.
+        - The `path` and `name` attributes must be defined in the instance for saving the file.
+
+        """
+        if self.details is None:
+            logger.error(f"The details attribute of the source {self.name} is not set.")
+            return []
+        if (
+            "url_archived" not in self.details
+            or self.details["url_archived"].isna().all()
+        ):
+            logger.error(
+                f"The url_archived attribute of source {self.name} is not set."
+            )
+            return []
+        if self.path is None:
+            logger.error(f"The path attribute of the source {self.name} is not set.")
+            return []
+
+        saved_paths: list[pathlib.Path | None] = []  # Explicit type annotation
+        for index, row in self.details.iterrows():
+            url_archived = row["url_archived"]
+            save_path = self._get_save_path(url_archived)
```

Comment:
Create the `save_path` from the `.details["title"]` by cleaning up the title of any undesired characters and then appending the file type.

This is also human-readable.

Current version has problems, if two entries with the same file extension exist for the same source.

Reply:
* Add a check to the loop if the path is already taken (don't overwrite files) 
* Add a check if the same `save_path` is already used by a different download in the loop (in case of same title or same file type)